src/effects/VST3/VST3Effect.cpp

```diff
@@ -0,0 +1,1198 @@
+            .Caption(XO("Error"))
+            .Parent(&dialogPlacement)
+      );
+      return;
+   }
+}
+
+void VST3Effect::ImportPresets()
+{
+   using namespace Steinberg;
+
+   auto path = SelectFile(FileNames::Operation::Presets,
+      XO("Load VST3 preset:"),
+      wxEmptyString,
+      wxEmptyString,
+      wxT(".vstpreset"),
+      {
+         { XO("VST3 preset file"), { wxT("vstpreset") }, true }
+      },
+      wxFD_OPEN | wxRESIZE_BORDER,
+      nullptr
+   );
+   if(path.empty())
+      return;
+
+   LoadPreset(path);
+}
+
+bool VST3Effect::HasOptions()
+{
+   return true;
+}
+
+void VST3Effect::ShowOptions()
+{
+   if(mEffectHost)
+   {
+      VST3OptionsDialog dlg(mParent, *mEffectHost, *this);
+      if (dlg.ShowModal())
+      {
+         ReloadUserOptions();
+      }
+   }
+}
+
+void VST3Effect::OnEffectWindowResize(wxSizeEvent& evt)
+{
+   using namespace Steinberg;
+
+   if(!mPlugView)
+      return;
+
+   const auto window = static_cast<wxWindow*>(evt.GetEventObject());
+   const auto windowSize = evt.GetSize();
+
+   {
+      //Workaround to prevent dialog window resize when
+      //plugin window reaches it's maximum size
+      auto root = wxGetTopLevelParent(window);
+      wxSize maxRootSize = root->GetMaxSize();
+
+      //remember the current dialog size as it's new maximum size
+      if(window->GetMaxWidth() != -1 && windowSize.GetWidth() >= window->GetMaxWidth())
+         maxRootSize.SetWidth(root->GetSize().GetWidth());
+      if(window->GetMaxHeight() != -1 && windowSize.GetHeight() >= window->GetMaxHeight())
+         maxRootSize.SetHeight(root->GetSize().GetHeight());
+      root->SetMaxSize(maxRootSize);
+   }
+   
+   ViewRect plugViewSize { 0, 0, windowSize.x, windowSize.y };
+   mPlugView->checkSizeConstraint(&plugViewSize);
+   mPlugView->onSize(&plugViewSize);
+}
+
+bool VST3Effect::LoadVSTUI(wxWindow* parent)
+{
+   using namespace Steinberg;
+   if(mEditController == nullptr)
+      return false;
+
+   bool useGUI { true };
+   GetConfig(*this, PluginSettings::Shared, wxT("Options"),
+            wxT("UseGUI"),
+            useGUI,
+            useGUI);
+   if(!useGUI)
+      return false;
+
+   if(const auto view = owned (mEditController->createView (Vst::ViewType::kEditor))) 
+   {  
+      parent->Bind(wxEVT_SIZE, &VST3Effect::OnEffectWindowResize, this);
+
+      ViewRect defaultSize;
+      if(view->getSize(&defaultSize) == kResultOk)
+      {
+         if(view->canResize() == Steinberg::kResultTrue)
+         {
+            ViewRect minSize {0, 0, parent->GetMinWidth(), parent->GetMinHeight()};
+            ViewRect maxSize {0, 0, 10000, 10000};
+            if(view->checkSizeConstraint(&minSize) != kResultOk)
+               minSize = defaultSize;
+
+            if(view->checkSizeConstraint(&maxSize) != kResultOk)
+               maxSize = defaultSize;
+
+            if(defaultSize.getWidth() < minSize.getWidth())
+               defaultSize.right = defaultSize.left + minSize.getWidth();
```

Comment:
I suspect off-by-one mistakes in these lines.

If the right coordinate of ViewRect is non-exclusive, which is how it works with wxSize, then it is left + width - 1.  Similarly for top, height, and bottom.

Reply:
Good.  The difference between the toolkits is surely worth commenting, as a possible source of confusion.